### ISSUE TITLE
Crashes during API call to /api/v1/session/show

### ISSUE THREAD

**Issue by alisoviy-mev** [vernemq/vernemq#619]
### Environment

- VerneMQ Version: 1.3.0
- OS: Ubuntu 16.04.4 LTS

We have a VerneMQ installed on AWS instance alongside with our application.
Every 10 seconds our app make API call to broker to get a list of sessions:
`http://localhost:8001/api/v1/session/show?--limit=200&--rowtimeout=10000`
Broker have up to 30 connection over TCP and up to 10 connections over WS.

As configuration we added this section to end of vernemq.conf:
```
########## Start ##########

listener.http.default=127.0.0.1:8001
listener.tcp.default=0.0.0.0:1883
listener.ws.default=0.0.0.0:1884

plugins.vmq_webhooks=on
vmq_webhooks.mywebhook1.hook=auth_on_register
vmq_webhooks.mywebhook1.endpoint=http://127.0.0.1:8080/authonregister
vmq_webhooks.mywebhook2.hook=auth_on_publish
vmq_webhooks.mywebhook2.endpoint=http://127.0.0.1:8080/authonpublish
vmq_webhooks.mywebhook3.hook=on_client_wakeup
vmq_webhooks.mywebhook3.endpoint=http://127.0.0.1:8080/onclientwakeup
vmq_webhooks.mywebhook4.hook=on_client_gone
vmq_webhooks.mywebhook4.endpoint=http://127.0.0.1:8080/onclientgone
vmq_webhooks.mywebhook5.hook=on_client_offline
vmq_webhooks.mywebhook5.endpoint=http://127.0.0.1:8080/onclientoffline
vmq_webhooks.mywebhook6.hook=auth_on_subscribe
vmq_webhooks.mywebhook6.endpoint=http://127.0.0.1:8080/authonsubscribe

graphite_enabled=on
graphite_host=127.0.0.1
graphite_port=2003
graphite_interval=10000
graphite_prefix=vernemq

########### End ###########
```

### Expected behavior

- Sessions list to be returned every time.

### Actual behaviour

Unfortunately I cannot provide any steps for reproduce, because we didn't found any specific actions before errors. And we hope that someone can provide us any info about how we can get additional info about the errors.

- Sometimes we had crash reports during API call (to get sessions list). It happens up to 15 times per day, and we cannot reproduce this errors.

crash.log:
```
2018-02-26 21:05:52 =ERROR REPORT====
** Generic server <0.21280.2> terminating 
** Last message in was {fetch,10}
** When Server state == {state,<0.21279.2>,undefined,33,[],[{'VerneMQ@127.0.0.1',<0.21281.2>}],1345323144,[]}
** Reason for termination == 
** {{normal,{gen_server,call,[<0.21281.2>,{fetch,desc,10},infinity]}},[{gen_server,call,3,[{file,"gen_server.erl"},{line,212}]},{vmq_ql_query,fetch,3,[{file,"/opt/vernemq/distdir/1.3.0/_build/default/lib/vmq_ql/src/vmq_ql_query.erl"},{line,55}]},{vmq_ql_query_mgr,'-fetch_results/2-fun-1-',4,[{file,"/opt/vernemq/distdir/1.3.0/_build/default/lib/vmq_ql/src/vmq_ql_query_mgr.erl"},{line,216}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_ql_query_mgr,fetch_results,2,[{file,"/opt/vernemq/distdir/1.3.0/_build/default/lib/vmq_ql/src/vmq_ql_query_mgr.erl"},{line,210}]},{vmq_ql_query_mgr,handle_call,3,[{file,"/opt/vernemq/distdir/1.3.0/_build/default/lib/vmq_ql/src/vmq_ql_query_mgr.erl"},{line,153}]},{gen_server,try_handle_call,4,[{file,"gen_server.erl"},{line,615}]},{gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,647}]}]}
2018-02-26 21:05:52 =CRASH REPORT====
  crasher:
    initial call: vmq_ql_query_mgr:init/1
    pid: <0.21280.2>
    registered_name: []
    exception exit: {{normal,{gen_server,call,[<0.21281.2>,{fetch,desc,10},infinity]}},[{gen_server,terminate,7,[{file,"gen_server.erl"},{line,812}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [<0.21279.2>]
    messages: []
    links: [<0.21279.2>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 712
  neighbours:
    neighbour: [{pid,<0.21279.2>},{registered_name,[]},{initial_call,{cowboy_protocol,init,4}},{current_function,{gen,do_call,4}},{ancestors,[]},{messages,[]},{links,[<0.406.0>,<0.21280.2>,#Port<0.24568>]},{dictionary,[]},{trap_exit,false},{status,waiting},{heap_size,6772},{stack_size,51},{reductions,2831}]
2018-02-26 21:05:52 =ERROR REPORT====
Ranch listener {{127,0,0,1},8001} had connection process started with cowboy_protocol:start_link/4 at <0.21279.2> exit with reason: {normal,{gen_server,call,[<0.21281.2>,{fetch,desc,10},infinity]}}
```
error.log:
```
2018-02-26 21:05:52.925 [error] <0.21280.2> gen_server <0.21280.2> terminated with reason: {normal,{gen_server,call,[<0.21281.2>,{fetch,desc,10},infinity]}} in gen_server:call/3 line 212
2018-02-26 21:05:52.925 [error] <0.21280.2> CRASH REPORT Process <0.21280.2> with 1 neighbours exited with reason: {normal,{gen_server,call,[<0.21281.2>,{fetch,desc,10},infinity]}} in gen_server:terminate/7 line 812
```

### Additional info

Sometimes when I tried to get sessions list by `vmq-admin` I had an error:
```
ubuntu@host:~$ sudo vmq-admin session show --limit=200 --rowtimeout=10000
RPC to 'VerneMQ@127.0.0.1' failed: {'EXIT',
                                    {function_clause,
                                     [{lists,zip,
                                       [[],"\t"],
                                       [{file,"lists.erl"},{line,387}]},
                                      {lists,zip,2,
                                       [{file,"lists.erl"},{line,387}]},
                                      {lists,zip,2,
                                       [{file,"lists.erl"},{line,387}]},
                                      {clique_table,'-max_widths/1-fun-0-',2,
                                       [{file,
                                         "/opt/vernemq/distdir/1.3.0/_build/default/lib/clique/src/clique_table.erl"},
                                        {line,227}]},
                                      {lists,foldl,3,
                                       [{file,"lists.erl"},{line,1263}]},
                                      {clique_table,get_field_widths,3,
                                       [{file,
                                         "/opt/vernemq/distdir/1.3.0/_build/default/lib/clique/src/clique_table.erl"},
                                        {line,103}]},
                                      {clique_table,autosize_create_table,3,
                                       [{file,
                                         "/opt/vernemq/distdir/1.3.0/_build/default/lib/clique/src/clique_table.erl"},
                                        {line,72}]},
                                      {clique_human_writer,write_table,1,
                                       [{file,
                                         "/opt/vernemq/distdir/1.3.0/_build/default/lib/clique/src/clique_human_writer.erl"},
                                        {line,67}]}]}}
```
And the same for: 
`sudo vmq-admin session show --limit=200`
`sudo vmq-admin session show --rowtimeout=10000`
But if I run this command without params or with limit that less than count of connected clients:
`sudo vmq-admin session show`
`sudo vmq-admin session show --limit=5 --rowtimeout=10000`
the list of sessions will be successfully returned.
But I'm not sure that these errors are related to errors from API calls.

**Comment by ioolkos** [comment]
@alisoviy-mev thanks a lot for your report! We'll look into that, I have the feeling it's a bug.

**Comment by ioolkos** [comment]
@alisoviy-mev haven't yet gotten to this.
@larshesel @dergraf ideas?

**Comment by larshesel** [comment]
I think I might have an idea - looking into it.

**Comment by alisoviy-mev** [comment]
@larshesel Can I provide some additional info?

**Comment by larshesel** [comment]
I was able to reproduce this a couple of times by repeating on of our tests cases 100 times (about 6/100 times it failed) and it seems to be a race-condition in the termination of one of the worker processes involved. I haven't figured out to fix this yet, but at least it's not reproducible to some degree. I'll see if I can figure this out next week.

**Comment by alisoviy-mev** [comment]
@larshesel Thank you very much for the answer! In this case there are some other questions: Is this error a critical one? I mean, should we worry about this or broker can restore its state in a safe way?

**Comment by larshesel** [comment]
It's not critical as it's only the query processes fetching session information that's crashing and it doesn't affect the state of the clients or broker in general.

**Comment by dergraf** [comment]
@alisoviy-mev we fixed a bug that's related to what you're seeing. However we weren't able to reproduce your original issue (only the second one). Please try with a new release (`make rel` on master) and see if you can still reproduce this issue.

**Comment by alisoviy-mev** [comment]
@dergraf Are you talking about issue with `vmq-admin` call? 
And as I understand, this fix wasn't released in 1.3.1 and I should build VerneMQ from master or download a nightly build. Right?

**Comment by larshesel** [comment]
Yeah, the fix is for the `vmq-admin session show` issue. You should be able to get the latest nightly build to test it out.

**Comment by alisoviy-mev** [comment]
@larshesel Ok, thanks! I'll try it asap.

**Comment by alisoviy-mev** [comment]
Still have a problem with the last nightly build (**vernemq-1.3.0-d586db43**):
```
ubuntu@host:~$ sudo vmq-admin session show --limit=200
RPC to 'VerneMQ@127.0.0.1' failed: {'EXIT',
                                    {function_clause,
                                     [{lists,zip,
                                       [[],"\t"],
                                       [{file,"lists.erl"},{line,387}]},
                                      {lists,zip,2,
                                       [{file,"lists.erl"},{line,387}]},
                                      {lists,zip,2,
                                       [{file,"lists.erl"},{line,387}]},
                                      {clique_table,'-max_widths/1-fun-0-',2,
                                       [{file,
                                         "/opt/vernemq/distdir/vernemq-1.3.0-d586db43/_build/default/lib/clique/src/clique_table.erl"},
                                        {line,227}]},
                                      {lists,foldl,3,
                                       [{file,"lists.erl"},{line,1263}]},
                                      {clique_table,get_field_widths,3,
                                       [{file,
                                         "/opt/vernemq/distdir/vernemq-1.3.0-d586db43/_build/default/lib/clique/src/clique_table.erl"},
                                        {line,103}]},
                                      {clique_table,autosize_create_table,3,
                                       [{file,
                                         "/opt/vernemq/distdir/vernemq-1.3.0-d586db43/_build/default/lib/clique/src/clique_table.erl"},
                                        {line,72}]},
                                      {clique_human_writer,write_table,1,
                                       [{file,
                                         "/opt/vernemq/distdir/vernemq-1.3.0-d586db43/_build/default/lib/clique/src/clique_human_writer.erl"},
                                        {line,67}]}]}}
```

**Comment by larshesel** [comment]
Ok, so this issue remains.

Can you share some more information about your setup? Are you running a single node (what's the output of `vmq-admin cluster show`?), how many clients are connected to each node? Are they websocket clients, plain tcp mqtt clients or anything else there? When you get the error using `vmq-admin session show`, do you see any other errors in the logs?

**Comment by alisoviy-mev** [comment]
We are running single node. There are about 20 clients: ~15 over TCP and ~5 over websockets. And I don't see any other messages in the logs.

**Comment by larshesel** [comment]
Hi @alisoviy-mev we were finally able to reproduce and understand the issue, there's a fix proposed in #961 

Thanks for your patience!

**Comment by larshesel** [comment]
Closed by #961
